command/validate.py

```diff
@@ -1,8 +1,239 @@
+        logger.info("Validating custom resource definitions.")
+        valid = True
+
+        crds = bundleData[self.crdKey]
+
+        for crd in crds:
+            if "metadata" in crd:
+                if "name" in crd["metadata"]:
+                    logger.info("Evaluating crd %s", crd["metadata"]["name"])
+                else:
+                    logger.error("crd metadata.name not defined.")
+                    valid = False
+            else:
+                logger.error("crd metadata not defined.")
+                valid = False
+
+            if "apiVersion" not in crd:
+                logger.error("crd apiVersion not defined.")
+                valid = False
+
+            if "spec" not in crd:
+                logger.error("crd spec not defined.")
+                valid = False
+
+        return valid
+
+    def _csv_validation(self, bundleData):
+        valid = True
+        logger.info("Validating cluster service versions.")
+
+        csvs = bundleData[self.csvKey]
+
+        for csv in csvs:
+            if "metadata" in csv:
+                if self._csv_metadata_validation(csv["metadata"]) is False:
+                    valid = False
+            else:
+                logger.error("csv metadata not defined.")
+                valid = False
+
+            if "apiVersion" not in csv:
+                logger.error("csv apiVersion not defined.")
+                valid = False
+
+            if "spec" in csv:
+                if self._csv_spec_validation(csv["spec"], bundleData) is False:
+                    valid = False
+            else:
+                logger.error("csv spec not defined.")
+                valid = False
+
+        return valid
+
+    def _csv_spec_validation(self, spec, bundleData):
+        valid = True
+
+        warnSpecList = ["displayName", "description", "icon", "version", "provider", "maturity"]
+        
+        for item in warnSpecList:
+            if item not in spec:
+                logger.warning("csv spec.%s not defined" % item)
+
+        if "installModes" not in spec:
+            logger.error("csv spec.installModes not defined" % item)
+            valid = False
+
+        if "install" not in spec:
+            logger.error("csv spec.install not defined" % item)
+            valid = False
+
+        if "customresourcedefinitions" in spec:
+            customresourcedefinitions = spec["customresourcedefinitions"]
+
+            crdList = []
+            for crd in bundleData[self.crdKey]:
+                try:
+                    name = crd["metadata"]["name"]
+                    crdList.append(name)
+                except KeyError:
+                    pass
+
+            if "owned" not in customresourcedefinitions:
+                logger.error("spec.customresourcedefinitions.owned not defined for csv")
+                valid = False
+            else:
+                for crd in customresourcedefinitions["owned"]:
+                    if "name" not in crd:
```

Comment:
Should we also check the versions of the CRD defined in the CSV and the CRDs provided in the bundle?

Reply:
@awgreene We can, but could we make that a separate feature / effort for later? It might be useful to have someone else work on this to make changes at a later point, and I think having every use case defined right now in the validator isn't necessary.

Reply:
I absolutely agree - this wasn't mean't to be a blocking issue. Let's add it later as an enhancement.